flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/api/ExpireSnapshots.java

```diff
@@ -0,0 +1,138 @@
+  }
+
+  public static class Builder extends MaintenanceTaskBuilder<ExpireSnapshots.Builder> {
+    private Duration maxSnapshotAge = null;
+    private Integer numSnapshots = null;
+    private int planningWorkerPoolSize = SystemConfigs.WORKER_THREAD_POOL_SIZE.value();
+    private int deleteBatchSize = DELETE_BATCH_SIZE_DEFAULT;
+    private int deleteParallelism = 1;
+
+    /**
+     * The snapshots older than this age will be removed.
+     *
+     * @param newMaxSnapshotAge of the snapshots to be removed
+     */
+    public Builder maxSnapshotAge(Duration newMaxSnapshotAge) {
+      this.maxSnapshotAge = newMaxSnapshotAge;
+      return this;
+    }
+
+    /**
+     * The minimum number of {@link Snapshot}s to retain. For more details description see {@link
+     * org.apache.iceberg.ExpireSnapshots#retainLast(int)}.
+     *
+     * @param newNumSnapshots number of snapshots to retain
+     */
+    public Builder retainLast(int newNumSnapshots) {
+      this.numSnapshots = newNumSnapshots;
+      return this;
+    }
+
+    /**
+     * The worker pool size used to calculate the files to delete.
+     *
+     * @param newPlanningWorkerPoolSize for planning files to delete
+     */
+    public Builder planningWorkerPoolSize(int newPlanningWorkerPoolSize) {
+      this.planningWorkerPoolSize = newPlanningWorkerPoolSize;
+      return this;
+    }
+
+    /**
+     * Size of the batch used to deleting the files.
+     *
+     * @param newDeleteBatchSize used for deleting
+     */
+    public Builder deleteBatchSize(int newDeleteBatchSize) {
+      this.deleteBatchSize = newDeleteBatchSize;
+      return this;
+    }
+
+    /**
+     * The number of subtasks which are doing the deletes.
+     *
+     * @param newDeleteParallelism used for deleting
+     */
+    public Builder deleteParallelism(int newDeleteParallelism) {
+      this.deleteParallelism = newDeleteParallelism;
+      return this;
+    }
+
+    @Override
+    DataStream<TaskResult> append(DataStream<Trigger> trigger) {
+      Preconditions.checkNotNull(tableLoader(), "TableLoader should not be null");
+
+      SingleOutputStreamOperator<TaskResult> result =
+          trigger
+              .process(
+                  new ExpireSnapshotsProcessor(
+                      tableLoader(),
+                      maxSnapshotAge == null ? null : maxSnapshotAge.toMillis(),
+                      numSnapshots,
+                      planningWorkerPoolSize))
+              .name(EXECUTOR_OPERATOR_NAME)
+              .uid(EXECUTOR_OPERATOR_NAME + uidSuffix())
+              .slotSharingGroup(slotSharingGroup())
+              .forceNonParallel();
+
+      result
+          .getSideOutput(ExpireSnapshotsProcessor.DELETE_STREAM)
+          .rebalance()
+          .transform(
+              DELETE_FILES_OPERATOR_NAME,
+              TypeInformation.of(Void.class),
+              new DeleteFilesProcessor(name(), tableLoader(), deleteBatchSize))
+          .name(DELETE_FILES_OPERATOR_NAME)
```

Comment:
It is probably better to include `uidSuffix` for operator name too. Otherwise, we may not be able to tell which is for which if we have multiple expire snapshots maintenance tasks in the same Flink job.

Reply:
Added the index of the builder to the operator for now.
I think it is better, as the same uid could be used for multiple Maintenance Tasks

Reply:
index is less friendly. users would be some like `Delete file 1` in Filnk UI. `uidSuffix` typically contains more meaningful id/name like Iceberg table name

Reply:
Index is guaranteed to be unique. If the user doesn't care about individually dropping the tasks state, then it would not configure the suffix independently for the tasks. If the suffix is set only on the TableManager level, then uids are not unique. BTW is that an issue? I have to check.

I really think that having multiple instances of the same task in the same TableMaintenance instance is more of an edge case. When I have seen the DAG on the UI, I have found that having long names would clutter the UI, especially when there are chained operators with long names.

Reply:
uid needs to be unique across the Flink job. `uidSuffix` is always combined with some operator name, like `DELETE_FILES_OPERATOR_NAME + uidSuffix()`. It seems that we are debating btw two options (1) `DELETE_FILES_OPERATOR_NAME + uidSuffix` (2) `DELETE_FILES_OPERATOR_NAME [index]`. 

>  multiple instances of the same task in the same TableMaintenance instance is more of an edge case

not sure about this comment and how is it related. but this scenario could happen if a Flink job writes to multiple Iceberg tables and maintenance actions (like expire snapshots) were appended for every table.

Reply:
On the UI these could be differentiated by the edges of the DAG.
Also having the uidSuffix in the name of the operator will cause the metrics name to change as well. I think that is undesired.

Reply:
Hmm, metric name/tag is an interesting question. 
1. if we don't include index or uidSuffix, then we can't distinguish metrics from multiple instances of the same maintenance task
2. uidSuffix. changing of uidSuffix would cause metric tag change
3. task index. the index could also change. adding a new task in the middle or removing a task from the list, which could cause change/shift of indexes.

Reply:
There is a 4th option: - allow setting an `id` for the job, which could default to the `index`, but I think this doesn't worth the complexity.

I would go with the 3rd option for now - as the number of tasks is normally not changing